### PR TITLE
Allow Formchecker to check more than 1 thing at a time

### PR DESCRIPTION
**Pull request by mawinter69**
The FormChecker is currently limiting the number of checks to one at a
time. While for most forms this is fine for the matrix auth and the
role-strategy this makes their config pages very slow to load. When 300
users are configured against an Active Directory it takes for me well
over 10 minutes until for each sid it is decided if it is a user or
group or doesn't exist.
Changing this needs to handled with care by those plugins. Browsers
limit the number of parallel connections one can open to the same host
usually with http1.1
But with http2 one can see big performance improvements. Using a
value of 50 and a role with 300 users the loading of user information
dropped from over 10 min to about 1 min for me.

### Proposed changelog entries

* Allow Formchecker to check more than one thing at a time

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6951"></a>

### CODE REVIEW

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 1]
```diff
@@ -198,6 +198,13 @@ var FormChecker = {
     // won't work.
     inProgress : 0,
 
+    // defines the maximum number of parallel checks to be run
+    // plugins that intend to to change this should be aware that when using http1.1 the
+    // browsers will usually throttle the number of connections and having a higher value
+    // can even have a negative impact. But with http2 enabled, this can be a great
+    // performance improvement
```
Comment (daniel-beck) [inline war/src/main/webapp/scripts/hudson-behavior.js:205]: Isn't this something core can take care of? Why would every plugin need to look at the HTTP version?

**Review by daniel-beck — COMMENTED** [review]
Could you provide a downstream PR (in pretty much any plugin, but ideally one demonstrating the benefit) that demonstrates use of this new property per the rule on new API?

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 1 above)
Comment (mawinter69) [reply war/src/main/webapp/scripts/hudson-behavior.js:205]: not sure how easily and reliably it can be checked within javascript if http2 is used or not.

**Comment by mawinter69** [comment]
Downstream PR:
https://github.com/jenkinsci/role-strategy-plugin/pull/228
This is just a PoC with the FormChecker copied over (with some not so nice code as one method called by FormChecker has changed between 2.303 as used by role-strategy and current weekly)

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 1 above)
Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:205]: this can be used apparently:
https://stackoverflow.com/a/46115491/4951015

**Comment by daniel-beck** [comment]
> This is just a PoC with the FormChecker copied over

You can create linked PRs that actually use new API by changing the dependency to the upstream PR's incrementals deployment. See https://github.com/jenkinsci/jenkins/pull/6951/checks?check_run_id=7621275049 for the current one for this PR.

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 1 above)
Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:205]: Still likely better handled in core if we can I think then it fixes it for everyone

Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:205]: I see you have, will check it later

Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:205]: I checked in safari and it works there too

Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:205]: the comment above this seems outdated now, as this isn't intended for plugins to change anymore from what I can tell? (given that it is unconditionally overridden on h2)

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 2]
```diff
@@ -242,6 +249,10 @@ var FormChecker = {
     }
 }
 
+if (performance.getEntriesByType("resource")[0].nextHopProtocol === "h2") {
```
Comment (daniel-beck) [inline war/src/main/webapp/scripts/hudson-behavior.js:252]: Is this defensive enough for this API? E.g., what if the list is empty? Can that happen?

**Comment by daniel-beck** [comment]
I really like where this is going :)

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 1 above)
Comment (mawinter69) [reply war/src/main/webapp/scripts/hudson-behavior.js]: I think with a value of 30 there would be no reason for matrix-auth or role-strategy plugins to overwrite that themselves.
But they would still be able to do so if they want.

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 2 above)
Comment (mawinter69) [reply war/src/main/webapp/scripts/hudson-behavior.js]: extracted the check to a method and safe guarded the array access and the field access

**Review by basil — APPROVED** [review]

**Review by daniel-beck — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by NotMyFault — APPROVED** [review]

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 1 above)
Comment (zbynek) [reply war/src/main/webapp/scripts/hudson-behavior.js]: > performance.getEntries()[0].nextHopProtocol worked with chrome but didn't work with firefox.

If I do `performance.getEntries()` in console of this PR, I get ~100 items, some of them `h2`, some of them not (both Chrome and FF). Could it be that in your case FF and Chrome just had a different resource at position 0?

**Pull request merged**